File: frontend/src/api/caseLive.ts
import { readApiResponse } from "./http";
import type { CaseData } from "../types";
import type { AgentRunState } from "../types/agent";
import type { CodeRunState } from "../types/agentCode";

export type CaseAgentExecMode = "position" | "code";

export interface LiveStepExecuteRequest {
  case_id?: number;
  case_name: string;
  script_content: string;
  steps: CaseData["steps"];
  commit_step_index: number;
  agent_mode: CaseAgentExecMode;
  run_id?: string;
  serial?: string | null;
  llm_provider?: string | null;
  enable_verifier?: boolean;
}

export interface LiveStepExecuteResponse {
  case: CaseData;
  run_id: string;
  agent_mode: CaseAgentExecMode;
  run: AgentRunState | CodeRunState;
  finished: boolean;
  message: string;
}

const API_BASE = "/api/cases/live";

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    headers: { "Content-Type": "application/json", ...options?.headers },
    ...options,
  });
  return readApiResponse<T>(response);
}

export const caseLiveApi = {
  executeStep: (payload: LiveStepExecuteRequest) =>
    request<LiveStepExecuteResponse>("/execute-step", {
      method: "POST",
      body: JSON.stringify({
        ...payload,
        serial: payload.serial ?? undefined,
        llm_provider: payload.llm_provider ?? undefined,
      }),
    }),
};

File: backend/app/schemas/case_live.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.case import CaseResponse, CaseStepCreate

AgentLiveMode = Literal["position", "code"]


class LiveStepExecuteRequest(BaseModel):
    """自然语言 Case 编写时提交单步并触发 AgentTest 执行。"""

    case_id: int | None = None
    case_name: str = "未命名 Case"
    script_content: str = ""
    steps: list[CaseStepCreate] = Field(default_factory=list)
    commit_step_index: int = Field(ge=0, description="本次确认执行的步骤下标（含）")
    agent_mode: AgentLiveMode = "position"
    run_id: str | None = None
    serial: str | None = None
    llm_provider: str | None = None
    enable_verifier: bool = False
    max_retries: int = Field(default=1, ge=0, le=3)


class LiveStepExecuteResponse(BaseModel):
    case: CaseResponse
    run_id: str
    agent_mode: AgentLiveMode
    run: dict[str, Any]
    finished: bool
    message: str
